File: leet-code/findMaxForm.js
/**
 * @param {string[]} strs
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var findMaxForm = function (strs, m, n) {
  let dp = new Array(m + 1).fill(0).map(() => new Array(n + 1).fill(0))

  for (let k = 0; k < strs.length; k++) {
    let zero = getZeroNum(strs[k])
    let one = getOneNum(strs[k])
    for (let i = m; i >= 0; i--) {
      for (let j = n; j >= 0; j--) {
        if (zero > i || one > j) {
          dp[i][j] = dp[i][j]
        } else {
          dp[i][j] = Math.max(1 + dp[i - zero][j - one], dp[i][j])
        }
      }
    }
  }
  return dp[m][n]
}

function getZeroNum (str) {
  let count = 0
  for (const char of str) {
    if (Number(char) === 0) {
      count++
    }
  }
  return count
}
function getOneNum (str) {
  let count = 0
  for (const char of str) {
    if (Number(char) === 1) {
      count++
    }
  }
  return count
}

console.log(findMaxForm(['10', '0001', '111001', '1', '0'], 5, 3))

File: leet-code/findMaxForm.go
package main

func findMaxForm(strs []string, m int, n int) int {
	dp := make([][]int, m+1)
	for i := range dp {
		temp := make([]int, n+1)
		dp[i] = temp
	}

	for _, s := range strs {
		zero, one := split(s)

		for i := m; i >= 0; i-- {
			for j := n; j >= 0; j-- {
				if zero > i || one > j {
					dp[i][j] = dp[i][j]
				} else {
					dp[i][j] = max(dp[i][j], 1+dp[i-zero][j-one])
				}
			}
		}
	}
	return dp[m][n]
}

func split(s string) (int, int) {
	zero, one := 0, 0
	for _, v := range s {
		if v == '0' {
			zero++
		} else if v == '1' {
			one++
		}
	}
	return zero, one
}
